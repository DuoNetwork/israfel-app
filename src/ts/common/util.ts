import { Constants, IUserOrder, Util as CommonUtil } from '@finbook/israfel-common';
import * as d3 from 'd3';
import moment from 'moment';
import * as CST from './constants';
import { ICustodianInfo } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = CommonUtil.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'just now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' min ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' hrs ago';
		else return 'long time ago';
	}

	public formatExpiry(isWeek: boolean) {
		return moment(CommonUtil.getExpiryTimestamp(isWeek)).format('YYYY-MM-DD HH:mm');
	}

	public formatTime(time: any) {
		return moment(time).format('MM-DD HH:mm:ss');
	}

	public formatTimeSecond(time: number) {
		return moment(time).format('YYYY-MM-DD HH:mm:ss');
	}

	public formatPriceShort(num: number) {
		if (num < 1) return d3.format('.3f')(num);
		return d3
			.format('.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatStrike(num: number) {
		return d3.format('.4f')(num);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatPercent(num: number) {
		return d3.format('.0%')(num);
	}

	public formatPercentAcc(num: number) {
		return d3.format('.2%')(num);
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatMaturity(maturity: number) {
		return maturity ? moment(maturity).format('YYYY-MM-DD HH:mm') : CST.TH_PERPETUAL;
	}

	public getOrderTitle(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order Placed';
		else if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_FILL) return 'Order Fully Filled';
			else if (order.status === Constants.DB_TERMINATE) return 'Order Expired';
			// order.status === Constants.DB_CONFIRMED
			else return 'Order Cancelled';
		else if (order.status === Constants.DB_MATCHING) return 'Order is Being Filled';
		// if (order.status === Constants.DB_PFILL)
		else return 'Order Partially Filled';
	}

	public getOrderFullDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		let baseDescription = `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}.`;
		if (order.type === Constants.DB_ADD)
			return baseDescription + ` Valid till ${this.formatMaturity(order.expiry)}`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_FILL)
			return baseDescription + ' Fully filled';

		if (order.fill || order.matching) {
			const totalFill =
				order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching);
			baseDescription += ` ${totalFill}(${this.formatPercent(
				totalFill / order.amount
			)}) filled.`;
		}

		if (order.matching && order.type !== Constants.DB_TERMINATE)
			baseDescription += 'Pending settlement';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_CONFIRMED)
			return baseDescription + ' Cancelled by user.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_TERMINATE)
			return baseDescription + ' Expired.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_BALANCE)
			return baseDescription + ` Cancelled due to insufficient balance.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_MATCHING)
			return baseDescription + ` Cancelled due to settlement error.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_RESET)
			return baseDescription + ` Cancelled due to custodian reset.`;

		return baseDescription;
	}

	public getOrderDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		return `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}. Total ${util.formatBalance(
			order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching)
		)} filled.`;
	}

	public getVersionDescription(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order placed.';

		if (order.type === Constants.DB_UPDATE) {
			let description = `Total ${this.formatBalance(order.fill + order.matching)} filled`;
			if (order.matching) description += ', pending settlement';
			return description;
		}

		if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_TERMINATE) return 'Expired';
			else if (order.status === Constants.DB_BALANCE)
				return 'Cancelled due to insufficent balance';
			else if (order.status === Constants.DB_MATCHING)
				return 'Cancelled due to settlement error';
			else if (order.status === Constants.DB_RESET) return 'Cancelled due to custodian reset';
			else if (order.status === Constants.DB_FILL) return 'Fully filled';
			// (order.status === Constants.DB_CONFIRMED)
			else return 'Cancelled by user';
		return 'Invalid order';
	}

	public getMaturityDescription(info?: ICustodianInfo) {
		return `This contract ${
			info && info.code.endsWith('PPT')
				? 'never expires'
				: 'expires in ' +
				(info
					? Math.floor(
							(info.states.maturity - CommonUtil.getUTCNowTimestamp()) / 86400000
						).toFixed(0)
					: 0) +
				' days'
		}.`;
	}

	public convertOrdersToCSV(orders: { [pair: string]: IUserOrder[] }): string {
		const headers = [
			CST.TH_PAIR,
			CST.TH_ORDER_HASH,
			CST.TH_VERSION,
			CST.TH_TIME_UTC,
			CST.TH_SIDE,
			CST.TH_PRICE,
			CST.TH_AMOUNT,
			CST.TH_FILL,
			CST.TH_FEE,
			CST.TH_EXPIRY_UTC,
			CST.TH_STATUS,
			CST.TH_TX_HASH
		];
		let output = '';
		headers.forEach(
			(header, index) =>
				(output += this.wrapCSVString(header) + (index < headers.length - 1 ? ',' : '\n'))
		);
		const ordersByHash: { [orderHash: string]: IUserOrder[] } = {};
		for (const pair in orders)
			orders[pair].forEach(order => {
				if (!ordersByHash[order.orderHash]) ordersByHash[order.orderHash] = [];
				ordersByHash[order.orderHash].push(order);
			});

		const rows: any[][] = [];
		for (const orderHash in ordersByHash) {
			ordersByHash[orderHash].sort((a, b) => a.currentSequence - b.currentSequence);
			ordersByHash[orderHash].forEach((order, index) => {
				const row: any[] = [];
				row.push(order.createdAt);
				row.push(order.pair.replace('|', '-')); // CST.TH_PAIR,
				row.push(order.orderHash); // CST.TH_ID,
				row.push(index + 1); // CST.TH_VERSION,
				row.push(
					moment.utc(order.updatedAt || order.createdAt).format('YYYY-MM-DD HH:mm:ss')
				); // CST.TH_TIME_UTC,
				row.push(order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL); // CST.TH_SIDE,
				row.push(order.price); // CST.TH_PRICE,
				row.push(order.amount); // CST.TH_AMOUNT,
				row.push(order.fill + order.matching); // CST.TH_FILL,
				row.push(order.fee + ' ' + order.feeAsset); // CST.TH_FEE,
				row.push(moment.utc(order.expiry).format('YYYY-MM-DD HH:mm:ss')); // CST.TH_EXPIRY,
				// CST.TH_STATUS,
				if (order.type === Constants.DB_TERMINATE)
					if (order.status === Constants.DB_BALANCE)
						row.push('Cancelled due to insufficent balance');
					else if (order.status === Constants.DB_MATCHING)
						row.push('Cancelled due to settlement error');
					else if (order.status === Constants.DB_RESET)
						row.push('Cancelled due to custodian reset');
					else if (order.status === Constants.DB_FILL) row.push('Fully filled');
					else if (order.status === Constants.DB_CONFIRMED) row.push('Cancelled by user');
					else row.push('Expired');
				else if (order.status === Constants.DB_MATCHING) row.push('Pending settlement');
				else row.push('Open');
				row.push(order.transactionHash || ''); // CST.TH_TX_HASH
				rows.push(row);
			});
		}

		rows.sort((a, b) => a[1].localeCompare(b[1]) || -a[0] + b[0] || -a[3] + b[3]);
		rows.forEach(row =>
			row.forEach((item, index) => {
				if (index)
					output +=
						this.wrapCSVString(item + '') + (index !== row.length - 1 ? ',' : '\n');
			})
		);

		return output;
	}

	public wrapCSVString(input: string): string {
		return ('' + input).indexOf(',') >= 0 ? '"' + input + '"' : input;
	}

	private getEtherScanLink() {
		return `https://${__ENV__ === Constants.DB_LIVE ? '' : 'kovan.'}etherscan.io`;
	}

	public getEtherScanTransactionLink(txHash: string) {
		return `${this.getEtherScanLink()}/tx/${txHash}`;
	}

	public getEtherScanTokenLink(tokenAddr: string, account: string = '') {
		return `${this.getEtherScanLink()}/token/${tokenAddr}${account ? `?a=${account}` : ''}`;
	}

	public getEtherScanAddressLink(account: string) {
		return `${this.getEtherScanLink()}/address/${account}`;
	}

	public fillZero(num: number) {
		return num < 10 ? `0${num}` : `${num}`;
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}
}

const util = new Util();
export default util;
