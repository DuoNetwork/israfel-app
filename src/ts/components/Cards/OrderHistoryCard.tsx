import { Table } from 'antd';
import { Icon, Popconfirm } from 'antd';
import moment from 'moment';
import * as React from 'react';
// import wsUtil from 'ts/common/wsUtil';
import * as CST from 'ts/common/constants';
import { IUserOrder } from 'ts/common/types';
import web3Util from 'ts/common/web3Util';
import wsUtil from 'ts/common/wsUtil';
// import util from '../../common/util';
import { SCard, SCardTitle } from './_styled';
import { STableWrapper } from './_styled';

const Column = Table.Column;

interface IProps {
	orderHistory: { [pair: string]: IUserOrder[] };
	account: string;
	display: boolean;
}

const parsePairRow: (pair: string) => any = pair => {
	return {
		key: pair,
		[CST.TH_PAIR]: pair,
		[CST.TH_SIDE]: '',
		[CST.TH_PX]: '',
		[CST.TH_AMT]: '',
		[CST.TH_BALANCE]: '',
		[CST.TH_FILL]: '',
		[CST.TH_MATCHING]: '',
		[CST.TH_FEE]: '',
		[CST.TH_EXPIRY]: '',
		[CST.TH_STATUS]: '',
		[CST.TH_ORDER_HASH]: '',
		[CST.TH_ACTIONS]: null,
		children: []
	};
};

const parseRow: (uo: IUserOrder, isParent: boolean, account: string, isCancel: boolean) => any = (
	uo,
	isParent,
	account,
	isCancel
) => {
	function handleClick(e: any, orderHash: string, pair: string, acc: string) {
		web3Util
			.web3PersonalSign(acc, CST.TERMINATE_SIGN_MSG + orderHash)
			.then(result => wsUtil.deleteOrder(pair, orderHash, result));
		console.log(e);
	}

	const row: { [key: string]: any } = {
		key: uo.currentSequence,
		[CST.TH_PAIR]: uo.pair,
		[CST.TH_SIDE]: uo.side === CST.DB_BID ? CST.TH_BUY : CST.TH_SELL,
		[CST.TH_PX]: uo.price,
		[CST.TH_AMT]: uo.amount,
		[CST.TH_BALANCE]: uo.balance,
		[CST.TH_FILL]: uo.fill,
		[CST.TH_MATCHING]: uo.matching,
		[CST.TH_FEE]: uo.fee + ' ' + uo.feeAsset,
		[CST.TH_EXPIRY]: moment(uo.expiry).format('YYYY-MM-DD HH:mm'),
		[CST.TH_STATUS]: uo.type + '|' + uo.status,
		[CST.TH_ORDER_HASH]: uo.orderHash,
		[CST.TH_ACTIONS]:
			isParent && isCancel ? (
				<Popconfirm
					title={CST.TT_DELETE_ORDER}
					icon={<Icon type="question-circle-o" style={{ color: 'red' }} />}
					onConfirm={e => handleClick(e, uo.orderHash, uo.pair, account)}
				>
					<a style={{ color: '#f5222d' }}>cancel</a>
				</Popconfirm>
			) : null
	};
	if (isParent) row.children = [];
	return row;
};

export default class OrderHistoryCard extends React.Component<IProps> {
	public render() {
		const { orderHistory, account, display } = this.props;
		const dataSource = [];
		for (const pair in orderHistory) {
			const pairHistory = orderHistory[pair];
			const pairRow = parsePairRow(pair);
			let parentRow = [];
			if (pairHistory.length) {
				pairHistory.sort(
					(a, b) =>
						-a.initialSequence + b.initialSequence ||
						-a.currentSequence + b.currentSequence
				);
				if (pairHistory[0].type === 'terminate')
					parentRow = parseRow(pairHistory[0], true, account, false);
				else parentRow = parseRow(pairHistory[0], true, account, true);
				for (let i = 1; i < pairHistory.length; i++) {
					const userOrder = pairHistory[i];
					if (userOrder.orderHash !== parentRow[CST.TH_ORDER_HASH]) {
						pairRow.children.push(parentRow);
						if (userOrder.type === 'terminate')
							parentRow = parseRow(userOrder, true, account, false);
						else parentRow = parseRow(userOrder, true, account, true);
					} else parentRow.children.push(parseRow(userOrder, false, account, false));
				}
				pairRow.children.push(parentRow);
				dataSource.push(pairRow);
			}
		}
		return (
			<SCard
				title={
					<SCardTitle>{(CST.TH_ORDER + ' ' + CST.TH_HISTORY).toUpperCase()}</SCardTitle>
				}
				className={'order-history-card ' + (display ? 'history-show' : '')}
				width="1080px"
				margin="0 10px 0 10px"
			>
				<STableWrapper>
					<Table dataSource={dataSource} pagination={false} style={{ width: '100%' }}>
						{[
							CST.TH_PAIR,
							CST.TH_SIDE,
							CST.TH_PX,
							CST.TH_AMT,
							CST.TH_BALANCE,
							CST.TH_FILL,
							CST.TH_MATCHING,
							CST.TH_FEE,
							CST.TH_EXPIRY,
							CST.TH_STATUS,
							CST.TH_ACTIONS
						].map(c => (
							<Column key={c} title={c} dataIndex={c} />
						))}
					</Table>
				</STableWrapper>
			</SCard>
		);
	}
}
