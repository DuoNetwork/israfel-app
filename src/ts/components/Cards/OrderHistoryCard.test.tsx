/**
 * @jest-environment jsdom
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import util from 'ts/common/util';
import OrderHistoryCard from './OrderHistoryCard';

describe('OrderHistoryCard Test', () => {
	describe('Test Snapshot', () => {
		const orderHistory = {
			'WETH|ETH': [
				{
					account: 'test',
					pair: 'test',
					orderHash: 'test',
					price: 123,
					amount: 123,
					balance: 123,
					fill: 123,
					side: 'ask',
					expiry: 123,
					createdAt: 123,
					updatedAt: 123,
					matching: 123,
					initialSequence: 123,
					currentSequence: 123,
					fee: 123,
					feeAsset: 'test',
					type: 'test',
					status: 'test',
					updatedBy: 'test',
					processed: true
				},
				{
					account: 'test',
					pair: 'test',
					orderHash: 'tesst123',
					price: 123,
					amount: 123,
					balance: 123,
					matching: 123,
					fill: 123,
					side: 'bid',
					expiry: 123,
					createdAt: 123,
					updatedAt: 123,
					initialSequence: 1234456,
					currentSequence: 1233,
					fee: 123,
					feeAsset: 'test',
					type: 'test',
					status: 'test',
					updatedBy: 'test',
					processed: true
				},
				{
					account: 'test',
					pair: 'test',
					orderHash: 'test',
					price: 123,
					matching: 111,
					amount: 123,
					balance: 123,
					fill: 123,
					side: 'test',
					expiry: 123,
					createdAt: 123,
					updatedAt: 123,
					initialSequence: 1234,
					currentSequence: 1234,
					fee: 123,
					feeAsset: 'test',
					type: 'test',
					status: 'test',
					updatedBy: 'test',
					processed: true
				}
			]
		};

		it('Test Snapshot', () => {
			util.formatTime = jest.fn(() => '1970-01-01 00:00:00');
			const wrapper = shallow(
				<OrderHistoryCard orderHistory={orderHistory} account={'test'} />
			);
			expect(wrapper).toMatchSnapshot();
		});
	});
});
