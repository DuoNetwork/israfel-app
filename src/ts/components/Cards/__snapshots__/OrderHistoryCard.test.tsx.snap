// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderHistoryCard Test Test Snapshot Test Snapshot 1`] = `
<Styled(Card)
  extra={
    <styled.button
      className="export-button"
      onClick={[Function]}
    >
      Export to CSV
    </styled.button>
  }
  margin="0 10px 50px 10px"
  title={
    <div
      className="ordercard-header"
      onClick={[Function]}
    >
      <div
        className="switch-button"
        style={
          Object {
            "borderLeft": "none",
            "marginLeft": 0,
            "opacity": 1,
            "paddingLeft": 0,
          }
        }
      >
        Open Orders
      </div>
      <div
        className="switch-button"
        style={
          Object {
            "opacity": 0.5,
          }
        }
      >
        Past Orders
      </div>
    </div>
  }
  width="740px"
>
  <styled.div>
    <Table
      bordered={false}
      className=""
      dataSource={
        Array [
          Object {
            "Actions": <Popconfirm
              icon={
                <Icon
                  style={
                    Object {
                      "color": "red",
                    }
                  }
                  type="question-circle-o"
                />
              }
              okType="primary"
              onConfirm={[Function]}
              placement="top"
              prefixCls="ant-popover"
              title="Are you sure to delete this order?"
              transitionName="zoom-big"
              trigger="click"
            >
              <a
                className="order-cancel"
              >
                Cancel
              </a>
            </Popconfirm>,
            "History": Array [
              Object {
                "account": "test",
                "amount": 123,
                "balance": 123,
                "createdAt": 123,
                "currentSequence": 123,
                "expiry": 123,
                "fee": 123,
                "feeAsset": "test",
                "fill": 123,
                "initialSequence": 123,
                "matching": 123,
                "orderHash": "test",
                "pair": "test",
                "price": 123,
                "processed": true,
                "side": "ask",
                "status": "test",
                "type": "test",
                "updatedAt": 123,
                "updatedBy": "test",
              },
              Object {
                "account": "test",
                "amount": 123,
                "balance": 123,
                "createdAt": 123,
                "currentSequence": 1234,
                "expiry": 123,
                "fee": 123,
                "feeAsset": "test",
                "fill": 123,
                "initialSequence": 1234,
                "matching": 111,
                "orderHash": "test",
                "pair": "test",
                "price": 123,
                "processed": true,
                "side": "test",
                "status": "test",
                "type": "test",
                "updatedAt": 123,
                "updatedBy": "test",
              },
            ],
            "Order": "Sell 123 test at 123 undefined per test. 246(100.000000%) filled.Pending settlement",
            "Time": "1970-01-01 00:00:00",
            "key": "test",
          },
          Object {
            "Actions": <Popconfirm
              icon={
                <Icon
                  style={
                    Object {
                      "color": "red",
                    }
                  }
                  type="question-circle-o"
                />
              }
              okType="primary"
              onConfirm={[Function]}
              placement="top"
              prefixCls="ant-popover"
              title="Are you sure to delete this order?"
              transitionName="zoom-big"
              trigger="click"
            >
              <a
                className="order-cancel"
              >
                Cancel
              </a>
            </Popconfirm>,
            "History": Array [
              Object {
                "account": "test",
                "amount": 123,
                "balance": 123,
                "createdAt": 123,
                "currentSequence": 1233,
                "expiry": 123,
                "fee": 123,
                "feeAsset": "test",
                "fill": 123,
                "initialSequence": 1234456,
                "matching": 123,
                "orderHash": "tesst123",
                "pair": "test",
                "price": 123,
                "processed": true,
                "side": "bid",
                "status": "test",
                "type": "test",
                "updatedAt": 123,
                "updatedBy": "test",
              },
            ],
            "Order": "Buy 123 test at 123 undefined per test. 246(100.000000%) filled.Pending settlement",
            "Time": "1970-01-01 00:00:00",
            "key": "tesst123",
          },
        ]
      }
      indentSize={20}
      loading={false}
      locale={Object {}}
      pagination={
        Object {
          "pageSize": 10,
          "pageSizeOptions": Array [
            "10",
            "20",
            "50",
          ],
          "showQuickJumper": true,
          "showSizeChanger": true,
          "showTotal": [Function],
          "size": "small",
        }
      }
      prefixCls="ant-table"
      rowKey="key"
      showHeader={true}
      size="default"
      style={
        Object {
          "width": "100%",
        }
      }
      useFixedHeader={false}
    >
      <Column
        dataIndex="Time"
        title="Time"
        width={180}
      />
      <Column
        dataIndex="Order"
        onCell={[Function]}
        title="Order"
      />
      <Column
        dataIndex="Actions"
        key="Actions"
        title=""
        width={100}
      />
    </Table>
  </styled.div>
  <OrderDetailCard
    handleClose={[Function]}
    orders={Array []}
  />
</Styled(Card)>
`;
